### PR TITLE
EvalState::realiseContext(): Allow access to the entire closure

### LINKED ISSUE #11030: Passing store paths into functions from other flakes causes errors in pure evaluation mode
**Affected releases (non-exhaustive):** Nix 2.20.6, Nix 2.21.2, Nix 2.22.1
**Last unaffected release:** Nix 2.19.4

It appears that recent versions of Nix have tightened pure evaluation mode. In the flake below, I pass a store path to a function from another flake, which used to work in older versions of Nix.

To reproduce, run `nix build` on the following flake:
```nix
{
  description = "Pass a store path to a flake";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs";
    nix-freeze-tree.url = "git+https://git.sr.ht/~jack/nix-freeze-tree?rev=60e09710d6062d54d063c8488337c14d9d302d58";
  };

  outputs = inputs: {
    packages.x86_64-linux.default =
      let
        pkgs = inputs.nixpkgs.legacyPackages.x86_64-linux;

        step1 = pkgs.runCommand "step1" {} ''
          mkdir -p $out
          echo '19 + 23' > $out/theanswer
        '';

        step2 = inputs.nix-freeze-tree.lib.x86_64-linux.freeze step1;
      in
      step2;
  };
}
```
You should see `error: access to absolute path '/nix/store/bhgy1rz0hi7b9anbnyr6zjgp0jh5dwrg-step1' is forbidden in pure evaluation mode (use '--impure' to override)`.

### PR DESCRIPTION
**Pull request by edolstra**
## Motivation

When doing IFD, we should allow access to the entire closure of the store paths built by the derivation, not just the top-level paths.

Fixes #11030.

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Hmm, makes me wonder if we ever want `allowPath` except for in the definition of `allowClosure`.

**Comment by Mic92** [comment]
> Hmm, makes me wonder if we ever want `allowPath` except for in the definition of `allowClosure`.

What about `builtins.toFile`?

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -347,6 +347,16 @@ void EvalState::allowPath(const StorePath & storePath)
         rootFS2->allowPrefix(CanonPath(store->toRealPath(storePath)));
 }
 
+void EvalState::allowClosure(const StorePath & storePath)
+{
+    if (!rootFS.dynamic_pointer_cast<AllowListSourceAccessor>()) return;
+
+    StorePathSet closure;
+    store->computeFSClosure(storePath, closure);
```
Comment (roberth) [inline src/libexpr/eval.cc:355]: Computing the closure from scratch could cause a performance problem.
`n` calls to `allowClosure` would be O(n * closure size), which may be significantly larger than O(total allowed closure size).
The latter can be achieved by computing the closure of paths that aren't in an allowed closure yet, i.e. checking the references recursively and skipping ones that have been of the closure of any preceding `allowClosure` call.

**Review by roberth — CHANGES_REQUESTED** [review]

File: src/libexpr/eval.cc (hunk 1 above)
Comment (edolstra) [reply src/libexpr/eval.cc:355]: The cost of the call to `computeFSClosure()` is likely to be trivial compared to the actual build of the derivation we're importing from.

**Comment by Mic92** [comment]
@mergify rebase

**Review by roberth — APPROVED** [review]

**Pull request merged**